### PR TITLE
Pre-surveys on My Professional Learning page

### PR DESCRIPTION
**Pull request by bencodeorg**
[PLC-800](https://codedotorg.atlassian.net/browse/PLC-800)

Adds a new button on the My Professional Learning page for teachers to access the pre-workshop survey. It is not displayed at all for workshops that have already started or do not have a pre-workshop survey, and is disabled until 10 days before the workshop. When disabled, it displays a message in a tooltip if a user hovers over it:

image

Three types of workshops have pre-surveys:
- 5-day summer workshops
- Academic year workshops
- Deep dive CSF workshops

The logic to get the URLs for each of these survey types is moved out of our mailers and into the enrollment model in this PR.

## Future work

We also gray out the "Print certificate" button on this page if a user hasn't actually attended a workshop, so now that there's a precedent for having a tooltip here I'd like to do the same there.

## Testing story

Adds unit tests on the Rails and JS sides to cover new behavior, with the exception of the tooltip. This PR also moves some repetitive test data into a factory using Rosie, and moves calls to `window.open` to `utils.windowOpen`, which caused some confusion during testing 😄.  

## Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/test/models/pd/enrollment_test.rb [hunk 1]
```diff
@@ -89,6 +89,27 @@ class Pd::EnrollmentTest < ActiveSupport::TestCase
...
+    csp_academic_year_workshop = build :csp_academic_year_workshop
+    csp_academic_year_workshop_enrollment = build :pd_enrollment, workshop: csp_academic_year_workshop
+
+    csf_deep_dive_workshop = build :csf_deep_dive_workshop
+    csf_201_workshop_enrollment = build :pd_enrollment, workshop: csf_deep_dive_workshop
+
+    csf_intro_workshop = build :csf_intro_workshop
+    csf_intro_workshop_enrollment = build :pd_enrollment, workshop: csf_intro_workshop
+
+    assert_equal "/pd/workshop_survey/day/0?enrollmentCode=#{csp_summer_workshop_enrollment.code}",
```
Comment (bencodeorg) [inline dashboard/test/models/pd/enrollment_test.rb:105]: This is a bit of a hack for the fact that `url_for` was generating HTTP URLs, and our helper `CDO.studio_url(relative_path` generated HTTPS.

**Comment by bencodeorg** [comment]
I could use someone's help thinking through if there are any unintended consequences of adding new fields to the workshop serializer here -- I think I've done this before without issue, but would love to walk through to make sure I understand.

**Review by molly-moen — APPROVED** [review]
LGTM! Not sure about implications of adding to the serializer so will leave that question for others

**Review by islemaster — APPROVED** [review]
Looks great!

The serializer additions look good in this case.  I think this serializer is only called explicitly by `Api::V1::Pd::WorkshopsController` in contexts where the additional info you added makes sense.  I was a little concerned that it might be modifying a CSV somewhere, but that doesn't seem to be the case.

**Pull request merged**